lib/dynamo-data.js

```diff
@@ -0,0 +1,174 @@
+import { promisify } from "node:util";
+import { inflate as zinflate } from "node:zlib";
+import log from "lambda-log";
+import { toEpochMilliseconds, toISODateString } from "./timestamp";
+
+const pinflate = promisify(zinflate);
+const inflate = async (buff) => JSON.parse(await pinflate(buff));
+const safeParseInt = (str) => {
+  if (str === undefined) {
+    return null;
+  } else {
+    const int = parseInt(str, 10);
+    return Number.isNaN(int) ? str : int;
+  }
+};
+
+/**
+ * Utility class for inserted dynamodb redirect + segments-downloaded data
+ */
+export default class DynamoData {
```

Comment:
This is a brand new class/abstraction.  The DynamoDB lambda code was so dang confusing, I needed an abstraction to make sense of it.

The constructor takes in things from `lib/dynamo.js` upsert call, and figures out which downloads/impressions for this request are now "counted".  And prevents double counting things.

Tests for this mostly existed elsewhere, just easier to unit-test things separately.